### DISCUSSION TITLE
How to Cancel Page Navigation

### DISCUSSION THREAD

**Discussion by Orgbrat** [CommunityToolkit/Maui#393] [Q&A]
So I have a Shell application with MVVM and it has a Items list page, NewItem page and a ItemDetail page. If you navigate to the NewItem page, enter some data but forget a field that is required then click Save there is code in the MVVM SaveItem function that pops up a MCT popup and tells the user the field that is required.

The issue is MAUI Shell does not wait for the Popup result but navigates back to the Items list page with the Popup still showing on top and the user has not even clicked his answer in the popup.

My question is how do you cancel the MAUI Shell Navigation process and stay on the NewItem page to correct the  missing data?

Orgbrat

**Comment by VladislavAntonyuk** [comment]
do you have a small repo? I guess you do not await the popup.

**Reply by Orgbrat** [reply]
Hi @VladislavAntonyuk 

I am awaiting the popup, but you can actually watch the Shell navigate back to Items list page underneath the displayed popup.
I really do not believe it is the Popop or the Toolkit that is the issue. I believe it to be the MAUI Shell framework.

If I add the OnNavigating method override to the AppShell.xaml.cs and look for the NewItem page I can cancel the navigation. The Popup is still there and still sitting on top of NewItem page. User can make choice on popup and all is as should be.  But this is an issue at application level. I need to be able to cancel navigation at the Page level. Is there not anyway to stop navigation in the page itself?

The actual app I am working on is rather large and private. So I will have to create a sample for you if needed.

**Reply by VladislavAntonyuk** [reply]
I would say you shouldn't initiate the navigation if your validation fails.
a small repro will be helpful

**Reply by pictos** [reply]
@Orgbrat I believe that you can use the `Shell.NavigatingEvent` to block the navigation, now this event has async support.
What you can do for a more generic handle is to have a `Dictionary<string,Func<Task<bool>>` to register and trigger the navigation handles. 
Here's an implementation [sample for you](https://github.com/MonkeyNights/maratona-shell/blob/aula-4-navegacao-parte-2/ShellMarin/ShellMarin/ShellMarin/NavigationService.cs#L44)
and you can find the [usage here](https://github.com/MonkeyNights/maratona-shell/blob/aula-4-navegacao-parte-2/ShellMarin/ShellMarin/ShellMarin/MainPage.xaml.cs#L40)

**Reply by pictos** [reply]
If my idea doesn't work, I would reforce the @VladislavAntonyuk 's request for a repo, that way we can take a good look on it

**Reply by Orgbrat** [reply]
Funny I have basically done what Pictos recommended but without the NavigationService. I added a flag to the application of stopNavigating and then added the OnNavigating overide in AppShell.cs. I set the application level flag from the viewModel and then in that Shell.OnNavigating function I check to see if I am popping the page and stopNavigating is true, if so I cancel the navigation. Then reset stopNavigating flag to false. Works perfectly but there should be a way to cancel a page navigation at the page level.

**Reply by pictos** [reply]
@Orgbrat I would say that is more about Shell than us. What you're saying makes sense and I believe that you can do it right now using AttachedProperties to control that... Maybe this can be suggested to be implemented in .NET MAUI, do you have in mind how you would like to use it in xaml or C#? Or even how the API should be implemented?

**Reply by Orgbrat** [reply]
@pictos I do agree with you that this is a Shell issue and not a Toolkit issue. There should be a page event where you could Cancel navigation if some condition was met on the page such as a validation error. But there is no such API available. As you showed me you have to go all the back to the Shell.OnNavigating with a flag value that is application scoped so the method can check the value and know to stop navigation.  I may open something in the .NET MAUI repo with this issue.

I thank you and @VladislavAntonyuk for your quick response and effort. You folks are the best...  :)

**Reply by Orgbrat** [reply]
@VladislavAntonyuk As a side note to your statement "I would say you shouldn't initiate the navigation if your validation fails.". In the SaveCommand method if there was page validation issues the code just showed the Popup and then just returned out of the SaveCommand. There was NO navigation initiated because the user needed to fix something on that page. But the Shell was automatically returning to the page that the NewItem page was called from which was what the issue was. Just thought I would share that with you.  I have coded a workaround  as I stated above. Still Thanks as you were of great assistance.

**Comment by CliffAgius** [comment]
Closed as answered
